docs: router HTTP response status codes (#4733)
apollographql/router: docs/source/errors.mdx

---
title: Errors
subtitle: Error and status codes returned by the Apollo Router
description: Error codes and response status codes returned by the Apollo Router
---

Learn about error codes and HTTP response status codes returned by the Apollo Router.

## Status codes

<PropertyList kind="errCodes">
<Property name="400" short="Bad request">

A problem when validating GraphQL or parsing an HTTP request.

</Property>
<Property name="401" short="Unauthorized">

> Sent an event to a non-existent subscription or used an invalid hash. Returned by either the authorization plugin to clients or the subscription endpoint.

Reviewer: @Geal I think I need more crispness on what you had meant when explaining this.  My **nuanced** take on what you had wrote was that this applied to ***subscription callback URL*** — which is _distinctly_ different from what Router clients would send to the Router.  In other words, two separate things:

1. When a client (e.g., Android, iOS) is making a request to the Route, if there is not a satisfactory answer from the JWT negotiation, then a 401 will be returned.
2. When using the subscription callback URL (i.e., when calling it from a subgraph implementation or pub/sub implementation; **not** a client-sourced request) if a:
   - Non-existent subscription identifier is used.
   - (I'm actually not sure what the difference is between an invalid hash and non-existent subscription — either way it's non-existent, I think?)

Reviewer 2: on 1. if there is an issue when verifying the JWT, or if no JWT is present and `require_authentication` is set to true, then a 401 is returned
on 2. yes, it is about the subscription callback URL. It can give the subgraph a key that is used to hash the subscription event. If the hash does not match the event, the router will return a 401 to the callback request. And there's another case in heartbeats where the router wil return a 401 if the id is not in the expected list, but I am not clear on that, maybe @bnjjj could give more context (or I'll dig in the code)

</Property>
<Property name="405" short="Method not allowed">

Used an unallowed HTTP method. Both mutations and subscriptions must use POST.

</Property>
<Property name="406" short="Not acceptable">

A query's Accept header doesn't match one of the supported content types: 
- `application/json`
- `application/graphql-response+json`
- `multipart/mixed;deferSpec=20220824`
- `multipart/mixed;subscriptionSpec=1.0`.

</Property>
<Property name="429" short="Too many requests">

Exceeded configured rate limits. See [client side traffic shaping](/configuration/traffic-shaping/#client-side-traffic-shaping).

</Property>
<Property name="500" short="Internal server error">

> An unexpected issue. Report this [possible bug](https://github.com/apollographql/router/issues/new?assignees=&labels=raised+by+user&projects=&template=bug_report.md&title=) to the router team.

Reviewer: It is also possible that the user has custom configured this HTTP status by throwing a particular Rhai error — I can't find this in the docs explicitly, but

- https://github.com/apollographql/router/pull/2097 introduced it.
- [This page mentions it](https://www.apollographql.com/docs/router/customizations/rhai/#:~:text=Headers-,Status%20codes,-Request%20context)

---
title: Errors
subtitle: Error and status codes returned by the Apollo Router

> description: Error codes and response status codes returned by the Apollo Router

Reviewer 3: Might as well include HTTP here for SEO:

```suggestion
description: Error codes and HTTP response status codes returned by the Apollo Router
```

---

Learn about error codes and HTTP response status codes returned by the Apollo Router.

## Status codes

<PropertyList kind="errCodes">
<Property name="400" short="Bad request">

> A request failed GraphQL validation, or a request failed to be parsed.

Reviewer 3: Definite articles seem more natural to me? Will just make the suggestion here rather than globally.

```suggestion
The request failed GraphQL validation or failed to be parsed.
```

</Property>
<Property name="401" short="Unauthorized">

> For a client making a request that requires authentication, a JWT failed to be verified.
> 
> For a non-client subscription endpoint calling a subscription callback URL, the router couldn't find a matching subscription identifier between its registered subscriptions and a subscription event.

Reviewer 3: ```suggestion
Requests may receive this response in two cases:

- For requests that requires authentication, the client's JWT failed verification.
- For a non-client subscription endpoint calling a subscription callback URL, the router couldn't find a matching subscription identifier between its registered subscriptions and a subscription event.
```

</Property>
<Property name="405" short="Method not allowed">

A request used an unallowed HTTP method.

<Note> 

Both mutations and subscriptions must use POST.

</Note>

</Property>
<Property name="406" short="Not acceptable">

> A query's HTTP `Accept` header didn't contain any of the router's supported mime-types: 

Reviewer 3: Is the switch from "request" --> "query" intentional here?

- `application/json`
- `application/graphql-response+json`
- `multipart/mixed;deferSpec=20220824`
- `multipart/mixed;subscriptionSpec=1.0`.

</Property>
<Property name="429" short="Too many requests">

Request traffic exceeded configured rate limits. See [client side traffic shaping](/configuration/traffic-shaping/#client-side-traffic-shaping).

</Property>
<Property name="500" short="Internal server error">

> The router encountered an unexpected issue. Report this [possible bug](https://github.com/apollographql/router/issues/new?assignees=&labels=raised+by+user&projects=&template=bug_report.md&title=) to the router team.

Reviewer 3: The link seems to fit better with the text "Report":

```suggestion
The router encountered an unexpected issue. [Report](https://github.com/apollographql/router/issues/new?assignees=&labels=raised+by+user&projects=&template=bug_report.md&title=) this possible bug to the router team.
```
